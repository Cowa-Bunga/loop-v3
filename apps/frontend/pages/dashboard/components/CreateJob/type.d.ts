interface IappCreateJobProps {
  handleClose: () => void
}
interface IappCreateJobState {
  data: {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    parcels: any[] | null
  }
  createForm: IForm
  parcelsForm: IForm
  parcels?: unknown[] | void
}
