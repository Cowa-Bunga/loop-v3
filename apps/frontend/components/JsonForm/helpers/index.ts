const makeInputId = (id: string, suffix?: string) => {
  const parts = id.split('/')

  if (suffix) {
    return `${parts[parts.length - 1]}-${suffix}`
  }

  return parts[parts.length - 1]
}

export { makeInputId }
