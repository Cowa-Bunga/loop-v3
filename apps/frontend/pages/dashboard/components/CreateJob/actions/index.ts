import { diff } from 'deep-object-diff'

export const Actions = (state, setState) => ({
  addEmptyParcel: () => {
    const { parcels } = state.data
    const updatedParcels = [...parcels, { qrCode: '', size: '' }]
    setState({ data: { parcels: updatedParcels } })
  },

  updateParcel: (index, parcel) => {
    const { parcels } = state.data
    parcels[index] = parcel
    setState({ data: { parcels } })
  },

  updateFormData: (data) => {
    if (Object.keys(diff(state.data, data)).length != 0) {
      setState({ data })
    }
  },

  getForms: async () => {
    const refs = ['create_task_form', 'parcels_form']
    const futures = refs.map((ref) => fetch(`/api/forms/${ref}`))

    try {
      const [create, parcels] = await Promise.all(futures)

      const createTaskForm = await create.json()
      const parcelsForm = await parcels.json()

      setState({
        data: createTaskForm.data,
        createForm: createTaskForm,
        parcelsForm: parcelsForm
      })
    } catch (e) {
      console.debug(e)
    }
  }
})
