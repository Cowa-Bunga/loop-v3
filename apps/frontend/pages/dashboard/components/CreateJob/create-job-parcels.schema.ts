export const createJobParcels = {
  schema: {
    type: 'object',
    properties: {
      qrCode: {
        type: 'string'
      },
      size: {
        type: 'string',
        enum: ['small', 'medium', 'large'] // enums should be localized keys
      }
    }
  },

  ui: {
    type: 'HorizontalLayout',
    elements: [
      {
        type: 'TextInputControl',
        scope: '#/properties/qrCode',
        i18n: 'create_task_form.parcels.qrCode'
      },
      {
        type: 'DropdownControl',
        scope: '#/properties/size',
        i18n: 'create_task_form.parcels.size'
      }
    ]
  },

  data: {
    qrCode: '',
    size: ''
  }
}
