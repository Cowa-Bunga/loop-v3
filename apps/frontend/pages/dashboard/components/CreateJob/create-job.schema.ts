export const createJob = {
  schema: {
    type: 'object',
    properties: {
      address: {
        type: 'string',
        options: {
          api_key: 'AIzaSyCgiluwpE3dNxGLL_iAPaV4SKZDTm_tpME',
          update_lat_long: true
        }
      },
      latitude: {
        type: 'integer'
      },
      longitude: {
        type: 'integer'
      },
      collectionTime: {
        type: 'string',
        format: 'date'
      },
      alcohol: {
        type: 'boolean'
      },
      taskNumber: {
        type: 'string'
      },
      customerName: {
        type: 'string'
      },
      mobileNo: {
        type: 'string'
      },
      deliveryTime: {
        type: 'string',
        format: 'date'
      },
      instructions: {
        type: 'string'
      },
      abandonFlow: {
        type: 'string',
        enum: ['sog', 'otp', 'img', 'qrCode', 'item'] // enums should be localized keys
      },
      deliveryFlow: {
        type: 'string',
        enum: ['sog', 'otp', 'img', 'qrCode', 'item'] // enums should be localized keys
      },
      taskType: {
        type: 'string',
        enum: ['delivery', 'collection'] // enums should be localized keys
      },
      parcels: {
        type: 'array',
        items: {
          type: 'object',
          properties: {
            qrCode: {
              type: 'string'
            },
            size: {
              type: 'string',
              enum: ['small', 'medium', 'large'] // enums should be localized keys
            }
          },
          required: ['qrCode', 'size']
        }
      }
    },
    required: [
      'collectionTime',
      'deliveryTime',
      'taskType',
      'customerName',
      'mobileNo',
      'taskNumber',
      'address',
      'latitude',
      'longitude'
    ]
  },

  ui: {
    type: 'Group',
    label: 'Create Task Form',
    i18n: 'create_task_form.title',
    elements: [
      {
        type: 'HorizontalLayout',
        elements: [
          // Left Side
          {
            type: 'VerticalLayout',
            elements: [
              {
                type: 'GooglePlacesControl',
                scope: '#/properties/address',
                i18n: 'create_task_form.address'
              },
              {
                type: 'TextInputControl',
                scope: '#/properties/latitude',
                i18n: 'create_task_form.latitude'
              },
              {
                type: 'TextInputControl',
                scope: '#/properties/longitude',
                i18n: 'create_task_form.longitude'
              },
              {
                type: 'DateTimeControl',
                scope: '#/properties/collectionTime',
                i18n: 'create_task_form.collectionTime'
              },
              {
                type: 'CheckboxControl',
                scope: '#/properties/alcohol',
                i18n: 'create_task_form.alcohol'
              },
              {
                type: 'MultiSelectControl',
                scope: '#/properties/abandonFlow',
                i18n: 'create_task_form.abandonFlow'
              }
            ]
          },
          // Right Side
          {
            type: 'VerticalLayout',
            elements: [
              {
                type: 'TextInputControl',
                scope: '#/properties/taskNumber',
                i18n: 'create_task_form.taskNumber'
              },
              {
                type: 'TextInputControl',
                scope: '#/properties/customerName',
                i18n: 'create_task_form.customerName'
              },
              {
                type: 'PhoneNumberControl',
                scope: '#/properties/mobileNo',
                i18n: 'create_task_form.mobileNo'
              },
              {
                type: 'DateTimeControl',
                scope: '#/properties/deliveryTime',
                i18n: 'create_task_form.deliveryTime'
              },
              {
                type: 'TextInputControl',
                scope: '#/properties/instructions',
                i18n: 'create_task_form.instructions'
              },
              {
                type: 'MultiSelectControl',
                scope: '#/properties/deliveryFlow',
                i18n: 'create_task_form.deliveryFlow'
              },
              {
                type: 'DropdownControl',
                scope: '#/properties/taskType',
                i18n: 'create_task_form.taskType'
              }
            ]
          }
        ]
      },
      // Bottom optional items like QR Codes and Items...
      {
        type: 'HorizontalLayout',
        elements: [
          {
            type: 'Control',
            scope: '#/properties/parcels',
            i18n: 'create_task_form.parcels',
            elements: [
              {
                type: 'Control',
                scope: '#/properties/parcels/properties/qrCode',
                i18n: 'create_task_form.qrCode'
              },
              {
                type: 'Control',
                scope: '#/properties/parcels/properties/size',
                i18n: 'create_task_form.size'
              }
            ]
          }
        ]
        // rule: {
        //   effect: 'SHOW',
        //   condition: {
        //     scope: '#',
        //     schema: {
        //       anyOf: [
        //         {
        //           properties: {
        //             deliveryFlow: {
        //               const: 'qrCode'
        //             }
        //           },
        //           required: ['deliveryFlow']
        //         },
        //         {
        //           properties: {
        //             abandonFlow: {
        //               const: 'qrCode'
        //             }
        //           },
        //           required: ['abandonFlow']
        //         }
        //       ]
        //     }
        //   }
        // }
      }
    ]
  },

  data: {
    address: '',
    latitude: 0,
    longitude: 0,
    collectionTime: '',
    alcohol: false,
    taskNumber: '',
    customerName: '',
    mobileNo: '',
    deliveryTime: '',
    instructions: '',
    abandonFlow: [],
    deliveryFlow: [],
    taskType: '',
    parcels: []
  }
}
