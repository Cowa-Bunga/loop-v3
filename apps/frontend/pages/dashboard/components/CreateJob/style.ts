export const ui = {
  container: {
    position: 'absolute',
    top: '10%',
    left: '50%',
    transform: 'translate(-50%, 0%)',
    width: '80%',
    height: '80%',
    overflow: 'scroll',
    backgroundColor: 'white'
  }
}
