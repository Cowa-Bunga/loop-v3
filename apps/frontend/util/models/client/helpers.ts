import { defaultsDeep } from 'lodash'
import { MODEL_CLIENT_EMPTY } from './model'

export const modelClientList = (
  data: { [s: string]: unknown } | ArrayLike<unknown>
): IClient[] => {
  return Object.entries(data)
    .map(modelClientData)
    .map((el: never) => el['1'])
    .sort((a: IClient, b: IClient) => a.name.localeCompare(b.name))
}

export const modelClientData = (data: unknown): IClient => {
  return defaultsDeep(data, MODEL_CLIENT_EMPTY)
}
