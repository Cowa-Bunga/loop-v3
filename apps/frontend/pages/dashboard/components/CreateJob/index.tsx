import { Box, CircularProgress, Modal } from '@mui/material'
import { JsonForm } from '@components'
import { useMergeState, useState, useTranslation } from '@hooks'
import ParcelsForm from '@pages/dashboard/components/CreateJob/components/ParcelsForm'
import { Actions } from './actions'
import { createJobFormLocalePathBuilder } from '@locale/locale-utils'
import { useEffect } from 'react'
import { IForm } from '@pages/api/forms/[reference]/type'

export interface CreateJobProps {
  handleClose: () => void
}

const style = {
  position: 'absolute',
  top: '10%',
  left: '50%',
  transform: 'translate(-50%, 0%)',
  width: '80%',
  height: '80%',
  overflow: 'scroll'
}

interface IState {
  data: any
  createForm: IForm
  parcelsForm: IForm
}

const CreateJob = ({ handleClose }: CreateJobProps) => {
  const [isLoading, setLoading] = useState(true)

  const [state, setState] = useMergeState<IState>({
    data: {},
    createForm: {} as IForm,
    parcelsForm: {} as IForm
  })
  const { addEmptyParcel, updateParcel, updateFormData, getForms } = Actions(
    state,
    setState
  )
  const { t } = useTranslation()
  const _t = (path: string) => t(createJobFormLocalePathBuilder(path))

  useEffect(() => {
    getForms().then(() => setLoading(false))

    return
  }, [])

  return (
    <Modal
      open
      onClose={handleClose}
      aria-labelledby={_t('aria.label')}
      aria-describedby={_t('aria.description')}
    >
      <Box sx={style}>
        {isLoading ? (
          <Box
            sx={style}
            display="flex"
            justifyContent="center"
            alignItems="center"
          >
            <CircularProgress />
          </Box>
        ) : (
          <>
            <JsonForm
              schema={state.createForm.schema}
              ui={state.createForm.ui}
              model={state.data}
              onChange={updateFormData}
            />

            <ParcelsForm
              ui={state.parcelsForm.ui}
              model={state.parcelsForm.data}
              schema={state.parcelsForm.schema}
              onChange={updateParcel}
              addParcel={addEmptyParcel}
              parcels={state.data.parcels}
            />
          </>
        )}
      </Box>
    </Modal>
  )
}

export default CreateJob
