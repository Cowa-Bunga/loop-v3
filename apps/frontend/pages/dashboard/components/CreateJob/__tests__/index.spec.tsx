import CreateJob from '../index'
import { renderWithProviders } from '@test/helpers'

jest.mock('react-i18next', () => ({
  ...jest.requireActual('react-i18next'),
  getI18n: () => ({
    resolvedLanguage: 'en'
  })
}))

jest.mock('@hooks', () => ({
  ...jest.requireActual('@hooks'),
  useTranslation: () => ({
    t: jest.fn().mockImplementation((key) => key)
  })
}))

jest.mock('./actions', () => ({
  Actions: jest.fn().mockImplementation(() => ({
    addEmptyParcel: jest.fn(),
    updateParcel: jest.fn(),
    updateFormData: jest.fn(),
    getForms: jest.fn().mockImplementation(() => Promise.resolve())
  }))
}))

describe('CreateJob', () => {
  it('should render open', () => {
    const { baseElement } = renderWithProviders(
      <CreateJob handleClose={console.log} isOpen={true} />
    )

    expect(baseElement).toMatchSnapshot()
  })

  it('should render closed', () => {
    const { baseElement } = renderWithProviders(
      <CreateJob handleClose={console.log} isOpen={false} />
    )
    expect(baseElement).toMatchSnapshot()
  })
})
