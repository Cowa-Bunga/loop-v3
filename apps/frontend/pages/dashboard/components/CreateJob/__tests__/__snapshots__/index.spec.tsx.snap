// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateJob should render closed 1`] = `
<body
  style=""
>
  <div />
</body>
`;

exports[`CreateJob should render open 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    aria-describedby="create_task_form.aria.description"
    aria-labelledby="create_task_form.aria.label"
    class="MuiModal-root css-79ws1d-MuiModal-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiModal-backdrop css-i9fmh8-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiBox-root css-m66zyu"
      tabindex="-1"
    >
      <div
        class="MuiBox-root css-byjdlg"
      >
        <span
          class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
          role="progressbar"
          style="width: 40px; height: 40px;"
        >
          <svg
            class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </span>
      </div>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
